Jobid/Resvid wrap it to zero after server restart multiple times consecutively

Bhagat-Rajput opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* Jobid/Resvid wrap it to zero after server restart multiple times consecutively 
* Steps to reproduce: 
   1. Submit a job and wait for the completion.
   2. Restart the server.
   3. Restart the server once again.
   4. Submit a job and this time will get the same jobid again (i.e. 0.servername)

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* PBS uses the global variable (sv_jobidnumber) to update the next jobidnumber in the database during shutdown.
 After server restart the variable intializes with zero again, now if server is restarted again (without submit a single job/resv) then its initialized value gets updated into the database. So next time whenever will start the PBS will get 0(Zero) jobid/resvid. 

#### Solution Description
* PBS will assign the updated jobidnumber value to the variable (sv_jobidnumber) at the time of server start/restart.

#### Testing logs/output
* Test logs: [ptl_pbs_trillion_jobid.txt](https://github.com/PBSPro/pbspro/files/2572311/ptl_pbs_trillion_jobid.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

suresh-thelkar commented:
It looks like the global variable svr_jobidnumber is not upto date in all the scenarios which is why the issue is happening. Since this is the root cause of the issue, I suggest to fix this part of the code instead of having the code to not save to database when svr_jobidnumber  is equal to zero. 

The other way to fix this is to eliminate the global variable svr_jobidnumber(we can make it a local variable in the appropriate place) and make sure that we always have upto date value in server.sv_qs.sv_jobidnumber. This is because when we save we always use this variable only.

suresh-thelkar commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -444,3 +444,27 @@ def test_jobid_duplication(self):
+        """
+        a = {'resources_available.ncpus': 15}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+        self.set_svr_sv_jobidnumber(0)
```
I don't think we need to reset svr_jobidnumber to zero in the database. In the system where this bug is present even without this step it is reproduced.

Bhagat-Rajput replied:
Its necessary because i am checking the job sequence as well.

arungrover replied:
I agree with @suresh-thelkar, Instead of setting it in DB, submit a job, get the job id and use that numeral to do further calculations

Bhagat-Rajput replied:
Addressed!

suresh-thelkar commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -444,3 +444,27 @@ def test_jobid_duplication(self):
+        self.submit_job(job_id='3')
+        self.submit_job(lower=1, upper=2, job_id='4[]')
+        self.submit_resv(resv_id='R5')
+        # Abruptly kill and start the server twice consecutively
```
Do we really need to check the case of abrupt killing of server as well ? In this case whatever value gets saved at the timing of abrupt kill of server, the same value gets recovered while server is getting restarted.

Bhagat-Rajput replied:
Yes it's important @suresh-thelkar. Now this functionality has been changed, after abruptly kill the server user will not get the next jobid/resvid.

Bhagat-Rajput commented:
> 
> 
> It looks like the global variable svr_jobidnumber is not upto date in all the scenarios which is why the issue is happening. Since this is the root cause of the issue, I suggest to fix this part of the code instead of having the code to not save to database when svr_jobidnumber is equal to zero.
> 
Thanks @suresh-thelkar for the review, svr_jobidnumber gets updated on a single job/resv run. If in a time frame PBS doesn't run a single job/resv then as per my knowledge no need to update the jobidnumber to the database because anyhow it would be the same.
But we can implement this behaviour as well.

> The other way to fix this is to eliminate the global variable svr_jobidnumber(we can make it a local variable in the appropriate place) and make sure that we always have upto date value in server.sv_qs.sv_jobidnumber. This is because when we save we always use this variable only.

If will go this way will loose the jobid/resvid because server saves its data from several other places as well.

suresh-thelkar commented:
> > It looks like the global variable svr_jobidnumber is not upto date in all the scenarios which is why the issue is happening. Since this is the root cause of the issue, I suggest to fix this part of the code instead of having the code to not save to database when svr_jobidnumber is equal to zero.
> 
> Thanks @suresh-thelkar for the review, svr_jobidnumber gets updated on a single job/resv run. If in a time frame PBS doesn't run a single job/resv then as per my knowledge no need to update the jobidnumber to the database because anyhow it would be the same.
> But we can implement this behaviour as well.
>> As discussed you can actually update the variable svr_jobidnumber as soon as svr_recov_db() gets called. This way before svr_shutdown() gets called in the said scenario it contains upto date value.
> 
> > The other way to fix this is to eliminate the global variable svr_jobidnumber(we can make it a local variable in the appropriate place) and make sure that we always have upto date value in server.sv_qs.sv_jobidnumber. This is because when we save we always use this variable only.
> 
> If will go this way will loose the jobid/resvid because server saves its data from several other places as well.
>> Since the save is happening in several other places this may not be trivial and hence option-1 is fine with me.

Bhagat-Rajput commented:
@suresh-thelkar I have addressed your comments please review it.

arungrover commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -444,3 +444,27 @@ def test_jobid_duplication(self):
+        server restart multiple times consecutively either gracefully/abruptly
+        """
+        a = {'resources_available.ncpus': 15}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
```
Why do you need to set node's resources available? You do not really care if the job is running or not.

Bhagat-Rajput replied:
Using a generic function self.submit_job() in which we were checking the R state, however have updated that as well.

Bhagat-Rajput commented:
Thanks @arungrover for finding the sequence window bug, I have addressed that as well along with this commit and added the testing scenario in PTL also. Please review it.

arungrover commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -144,7 +144,7 @@ def stop_and_restart_svr(self, restart_type):
 
     def submit_job(self, sleep=10, lower=0,
-                   upper=0, job_id=None, job_msg=None):
+                   upper=0, job_id=None, job_msg=None, status=None):
```
Since the function only checks for 'R' state, I'd suggest you rename to status to something like 'running' or 'check_running' and make it a boolean with a default value as 'False'.

Bhagat-Rajput replied:
Addressed!

arungrover commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -178,15 +181,15 @@ def submit_job(self, sleep=10, lower=0,
+                    self.server.expect(JOB, {'job_state': 'B'}, id=jid)
+                    self.server.expect(
+                        JOB,
+                        {'job_state=R': '%d' % (total_jobs)},
```
Can you not write total_jobs to the dictionary instead of substituting it as a string?

Bhagat-Rajput replied:
Addressed!

Bhagat-Rajput commented on src/server/req_quejob.c:
```diff
@@ -3236,7 +3236,7 @@ long long get_next_svr_sequence_id(void)
 	++svr_sequence_window_count;
 	ret_svr_sequence_id = svr_jobidnumber;
 	/* sequence window count is more than 1000, reset back to zero*/
-	if (svr_sequence_window_count > SEQ_WIN_INCR) {
```
As per my understanding, we can't put this line in the end because at line 3243 we are checking the next jobid that should not be greater than the svr_max_job_sequence_id attribute. If  greater then need to wrap it to 0 from here itself.
Will not loose the jobid because we are assigning first (at line 3223) and then increasing the jobid later.
I think previously also PBS saves the advance jobid in the database and written a testcase as well to check this scenario.

arungrover replied:
Ok, in that case change the operator in te check to be '>=' just to be on a safer side

Bhagat-Rajput replied:
Done!

Bhagat-Rajput commented:
Thanks @arungrover and @vstumpf , addressed all the comments. Please review it.

Bhagat-Rajput commented:
Rebased and squashed, ready to merge.

agrawalravi90 commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -165,28 +165,31 @@ def submit_job(self, sleep=10, lower=0,
             j.set_attributes({ATTR_J: '%d-%d' % (lower, upper)})
             arr_flag = True
-            total_jobs = upper - lower + 1
+            arr_jobs = {'total_jobs': '%d' % (upper - lower + 1)}
```
why did you turn this into a dict? looks like you only ever store one key-val pair into it

agrawalravi90 replied:
It's always wise to do whatever Arun tells you to :) but I think you misunderstood what he meant to say. I think he was just asking you to pass total_jobs directly to the attribute dictionary which self.server.expect() takes as its second argument, instead of passing it as a string. i.e - he asked you to do
self.server.expect(JOB, {'job_state=R': total_jobs}, ..)
instead of 
{'job_state=R': '%d' % (total_jobs)}, ..)
which you had previously.

Bhagat-Rajput replied:
Addressed!

agrawalravi90 commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -165,28 +165,31 @@ def submit_job(self, sleep=10, lower=0,
+                        count=True, id=jid, extend='t')
             else:
-                self.server.expect(JOB, {'job_state': 'R'}, id=jid)
-                if sleep == 1:
```
not sure what the sleep == 1 check was for, but can you please comment on why you removed it?

agrawalravi90 commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -144,7 +144,7 @@ def stop_and_restart_svr(self, restart_type):
 
     def submit_job(self, sleep=10, lower=0,
-                   upper=0, job_id=None, job_msg=None):
+                   upper=0, job_id=None, job_msg=None, running=False):
```
how about 'verify' instead of 'running' ?

agrawalravi90 replied:
Also, you call this with running=True everywhere, so why did you decide to add this argument? Earlier it was always going to check that the job was in the 'R' state, and that's exactly what it will do even with this new argument because you call it with 'True' everywhere, right?

Bhagat-Rajput replied:
No @agrawalravi90, only in one testcase i am checking the running state of jobs.

agrawalravi90 replied:
my bad, thanks for clarifying

agrawalravi90 commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -444,3 +451,34 @@ def test_jobid_duplication(self):
+        # Gracefully stop and start the server twice consecutively
+        self.stop_and_restart_svr('normal')
+        self.stop_and_restart_svr('normal')
+        self.submit_job(job_id='%s' % str(curr_id + 4))
```
you aren't really checking that the job id wrapped to 0 or not, wouldn't it be a simpler and more targeted test if you just submitted a single job after the server restarts and check that its id is not 0 ?

Bhagat-Rajput replied:
self.submit_job() is a generic function and using throughout the whole testsuite in which we are checking the exact jobid numbers in other test cases.
So that's why used the same in this test also and if it will wrap to zero after server restart then testcase will fail anyway.
Checking the array jobs and reservation as well because they also gets the id from the same function.

agrawalravi90 commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -444,3 +451,34 @@ def test_jobid_duplication(self):
+        self.submit_resv(resv_id='R%s' % str(curr_id + 2))
+
+    def tearDown(self):
+        self.server.cleanup_jobs(extend='force')
```
forcefully deleting the jobs doesn't really delete the jobs on the mom, which means that if another test is run immediately following this test, the next test might run with all of those jobs lurking around on the mom. So, I think you should remove the 'force' flag and delete jobs properly in tearDown.

Bhagat-Rajput replied:
Thanks @agrawalravi90 for the valuable suggestion, implemented.

agrawalravi90 commented on test/tests/functional/pbs_trillion_jobid.py:
```diff
@@ -480,5 +480,5 @@ def test_jobid_resvid_after_multiple_restart(self):
 
     def tearDown(self):
-        self.server.cleanup_jobs(extend='force')
+        #self.server.cleanup_jobs()
```
I think you forgot to un-comment this line

Bhagat-Rajput replied:
Done!

Bhagat-Rajput commented:
I've attached the latest test logs below: 
[ptl_pbs_trillion_jobid.txt](https://github.com/PBSPro/pbspro/files/2585558/ptl_pbs_trillion_jobid.txt)

agrawalravi90 commented:
Looks good Bhagat, please allow Arun to look at the latest changes once, then either of us can take this forward.

arungrover commented:
Change looks good to me

agrawalravi90 commented:
Please rebase for merge bhagat

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arun Grover <notifications@github.com>
Sent: Thursday, November 15, 2018 8:38:34 PM
To: PBSPro/pbspro
Cc: Ravi Agrawal; Mention
Subject: Re: [PBSPro/pbspro] Jobid/Resvid wrap it to zero after server restart multiple times consecutively (#878)


Change looks good to me

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/PBSPro/pbspro/pull/878#issuecomment-439072535>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEXOm4bSCD_qzgwFx9upwVl85zyDmiLxks5uvYNygaJpZM4YZ0rP>.

Bhagat-Rajput commented:
Squashed and Rebased, please take this further. Thanks

arungrover commented:
Appveyor failed initially, I have restarted the build

Bhagat-Rajput commented:
Thanks @arungrover, it's passed now -:)

agrawalravi90 commented:
Sorry Bhagat, I noticed your comment just now, I had already merged a PR before that. Can you please rebase again? I'll merge your PR before anybody else's. Thanks.

Bhagat-Rajput commented:
It's fine @agrawalravi90. I've rebased it. Thanks!
